### ISSUE TITLE
Anyway to clear/reset the static shape to None?

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#5680]
`Tensor.set_shape()` doesn't set shape but merge shape. If I don't want the statically-inferred shape, could there be a method to clear it?

The use case is as following: I have:
```python
x = tf.random_uniform([128,10])
# other operations involving x
```
in the graph, but I want to be able to run it (in inference time) with `feed_dict={x: value}` where `value` is a tensor of shape [1, 10]. It would work perfectly if I could somehow set the shape of x to [None, 10].
Currently I'm using `tf.shape()` as argument of `random_uniform` to avoid shape inference and `set_shape` later, which is not a very nice solution.

**Comment by mrry** [comment]
There's deliberately no way to make a tensor's shape *less* specific. This makes it safe for TensorFlow to use the current static shape of a tensor in later shape inference, and also to specialize the graph for a specific shape (see e.g. the implementation of `tf.shape()`, which returns a constant if the shape is known).

However, your feeding use case is supported, using the [`tf.placeholder_with_default()`](https://www.tensorflow.org/versions/r0.11/api_docs/python/io_ops.html#placeholder_with_default) op, which lets you define a feed point with both (i) a default value when you don't feed it, and (ii) a less specific static shape than the default value. You could use it in your case as follows:

```python
x_default = tf.random_uniform([128,10])
x = tf.placeholder_with_default(x_default, shape=[None, 10])
# other operations involving x
```

**Comment by ppwwyyxx** [comment]
Thanks! `placeholder_with_default` works for my case.

**Comment by ZhimingZhou** [comment]
HI, I'm also facing a similar issue that requires to modify the static shape to None. But it cannot be handled by placeholder_with_default. I'm loading a pre-trained model/graph that has a static shape. 

The thing is that I'm using the Inception Score, which is a very important evaluation metric in generative models, especially the GANs community.  

It involves the pre-trained model http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz. But it has static batch size 1, which makes the inference extremely slow. It was used to be handled by the following code, i.e. changing the tensor shape via tensor._shape = new_shape.

       ```
 ops = self.inception_graph.get_operations()
        for op_idx, op in enumerate(ops):
            for o in op.outputs:
                shape = o.get_shape()
                shape = [s.value for s in shape]
                new_shape = []
                for j, s in enumerate(shape):
                    if s == 1 and j == 0:
                        new_shape.append(None)
                    else:
                        new_shape.append(s)
                # o._shape = tf.TensorShape(new_shape)
```
The above operation works for tensorflow 1.5 and earlier version. But for the latter version such as tensorflow 1.9, o._shape = tf.TensorShape(new_shape) is no longer allowed, and set_shape only allows to update shape, but no support modifying known to None.

Updating shape seems necessary here.  So I'm looking forward to a update_shape operation. I understand that "there's deliberately no way to make a tensor's shape less-specific." But any idea to solve the above-mentioned problem, any help? The current version forces me to use tensorflow up to version 1.5.

Thanks a lot!

**Comment by mrry** [comment]
You can use the `input_map` argument in `tf.import_graph_def()` to rebind the input tensor to a `tf.placeholder()` created in the target graph.
